src/cryptonote_core/blockchain.cpp

```diff
@@ -3740,20 +3769,110 @@ bool Blockchain::check_blockchain_pruning()
+
+  uint64_t short_term_constraint = long_term_effective_median_block_weight + long_term_effective_median_block_weight * 2 / 5;
+  uint64_t long_term_block_weight = std::min<uint64_t>(block_weight, short_term_constraint);
+
+  return long_term_block_weight;
+}
+//------------------------------------------------------------------
+bool Blockchain::update_next_cumulative_weight_limit(uint64_t *long_term_effective_median_block_weight)
 {
-  uint64_t full_reward_zone = get_min_block_weight(get_current_hard_fork_version());
+  PERF_TIMER(update_next_cumulative_weight_limit);
 
   LOG_PRINT_L3("Blockchain::" << __func__);
-  std::vector<size_t> weights;
-  get_last_n_blocks_weights(weights, CRYPTONOTE_REWARD_BLOCKS_WINDOW);
 
-  uint64_t median = epee::misc_utils::median(weights);
-  m_current_block_cumul_weight_median = median;
-  if(median <= full_reward_zone)
-    median = full_reward_zone;
+  uint64_t local_long_term_effective_median_block_weight;
+  if (!long_term_effective_median_block_weight)
+    long_term_effective_median_block_weight = &local_long_term_effective_median_block_weight;
+
+  // when we reach this, the last hf version is not yet written to the db
+  const uint64_t db_height = m_db->height();
+  const uint8_t hf_version = (m_long_term_block_weights_height + 1 < m_db->height()) ? m_db->get_hard_fork_version(db_height - 1) : get_current_hard_fork_version();
+  uint64_t full_reward_zone = get_min_block_weight(hf_version);
+  uint64_t long_term_block_weight;
+
+  // recalc, or reorg ? pop the last one(s) and recompute
+  uint64_t pop = 0;
+  if (db_height && db_height <= m_long_term_block_weights_height)
+  {
+    pop = m_long_term_block_weights_height - db_height + 1;
+    while (pop--)
+      pop_from_long_term_block_weights();
+  }
+
+  if (hf_version < HF_VERSION_LONG_TERM_BLOCK_WEIGHT)
+  {
+    std::vector<uint64_t> weights;
+    get_last_n_blocks_weights(weights, CRYPTONOTE_REWARD_BLOCKS_WINDOW);
+    m_current_block_cumul_weight_median = epee::misc_utils::median(weights);
+    long_term_block_weight = weights.back();
+  }
+  else
+  {
+    const uint64_t block_weight = m_db->get_block_weight(db_height - 1);
+
+    std::vector<uint64_t> weights(m_long_term_block_weights.begin(), m_long_term_block_weights.end());
+    uint64_t long_term_median = epee::misc_utils::median(weights);
+    *long_term_effective_median_block_weight = std::max<uint64_t>(CRYPTONOTE_BLOCK_GRANTED_FULL_REWARD_ZONE_V5, long_term_median);
+
+    uint64_t short_term_constraint = *long_term_effective_median_block_weight + *long_term_effective_median_block_weight * 2 / 5;
+    long_term_block_weight = std::min<uint64_t>(block_weight, short_term_constraint);
+
+    weights = std::vector<uint64_t>(m_long_term_block_weights.begin(), m_long_term_block_weights.end());
+    if (weights.empty())
+      weights.resize(1);
+    weights[0] = long_term_block_weight;
+    long_term_median = epee::misc_utils::median(weights);
```

Comment:
Why is the computation of long term median performed twice? What does `weights[0] = long_term_block_weight;` mean?

Reply:
It's done once over the last N, then once over the last N-1 plus the current one. weights[0] = long_term_block_weight just removes the oldest element, and adds the newest one.